sycl/include/sycl/detail/spirv.hpp

```diff
@@ -775,217 +803,322 @@ inline uint32_t membermask() {
+                detail::is_vec<T>::value) {
+    // Temporary work-around due to a bug in IGC.
+    // TODO: Remove when IGC bug is fixed.
+    T result;
+    for (int s = 0; s < x.size(); ++s)
+      result[s] = ShuffleXor(g, x[s], mask);
+    return result;
+  } else if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                           GroupT>) {
+    // Since the masks are relative to the groups, we could either try to adjust
+    // the mask or simply do the xor ourselves. Latter option is efficient,
+    // general, and simple so we go with that.
+    id<1> TargetLocalId = g.get_local_id() ^ mask;
+    uint32_t TargetId = MapShuffleID(g, TargetLocalId);
+    return __spirv_GroupNonUniformShuffle(group_scope<GroupT>::value,
+                                          convertToOpenCLType(x), TargetId);
+  } else {
+    // Subgroup.
+    return __spirv_SubgroupShuffleXorINTEL(convertToOpenCLType(x),
+                                           static_cast<uint32_t>(mask.get(0)));
+  }
 #else
-  return __nvvm_shfl_sync_bfly_i32(membermask(), x, local_id.get(0), 0x1f);
+  if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                    GroupT>) {
+    return __nvvm_shfl_sync_bfly_i32(detail::ExtractMask(detail::GetMask(g))[0],
+                                     x, static_cast<uint32_t>(mask.get(0)),
+                                     0x1f);
+  } else {
+    return __nvvm_shfl_sync_bfly_i32(membermask(), x,
+                                     static_cast<uint32_t>(mask.get(0)), 0x1f);
+  }
 #endif
 }
 
-template <typename T>
-EnableIfNativeShuffle<T> SubgroupShuffleDown(T x, uint32_t delta) {
+template <typename GroupT, typename T>
+EnableIfNativeShuffle<T> ShuffleDown(GroupT g, T x, uint32_t delta) {
 #ifndef __NVPTX__
-  return __spirv_SubgroupShuffleDownINTEL(convertToOpenCLType(x),
-                                          convertToOpenCLType(x), delta);
-#else
-  return __nvvm_shfl_sync_down_i32(membermask(), x, delta, 0x1f);
+  if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                    GroupT> &&
+                detail::is_vec<T>::value) {
+    // Temporary work-around due to a bug in IGC.
+    // TODO: Remove when IGC bug is fixed.
+    T result;
+    for (int s = 0; s < x.size(); ++s)
+      result[s] = ShuffleDown(g, x[s], delta);
+    return result;
+  } else if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                           GroupT>) {
+    // To avoid overflow we adjust the delta down by the size of the group, if
+    // it is larger than the size of the group. Since the local ID cannot be
+    // larger than the size of the group, it cannot overflow when we add them
+    // prior to wrapping the target ID.
+    id<1> TargetLocalId = g.get_local_id();
+    size_t ItemsInGroup = g.get_local_linear_range();
+    uint32_t AdjustedDelta =
+        delta > ItemsInGroup ? delta - ItemsInGroup : delta;
```

Comment:
Should this be `>=`?  Shifting by `ItemsInGroup` will also give you an index outside of the group, I think?